Session/2 API 天気予報を取得し、表示する

KokiIto-45 opened the pull request:
YumemiWeatherの擬似的な APIを利用し、天気予報を取得→UIImageViewに該当する天気の画像を表示するよう機能追加しました。

KokiIto-45 committed:
session 1
AutoLayoutで初期画面を作成

KokiIto-45 committed:
session 2 APIの利用

YumemiWeatherを利用して天候状況を取得して画像として表示する機能を追加

KokiIto-45 committed:
session/1 AutoLayout

UILabelのwidthの基準をsuperViewからUIImageViewに変更

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -6,13 +6,40 @@
+    
+    @IBOutlet weak var weatherImageView: UIImageView!
+    
+    private var weather:String = ""
```
> weather:String

ここの「:」の後のスペースの有無（Swiftでは基本的に`var hoge: Hoge = Hoge()`の形）とか、まえ秋月さんにも言われてたけど気になってくるね〜

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -6,13 +6,40 @@
+    
+    private func setWeatherImage() {
+        weather = YumemiWeather.fetchWeatherCondition()
+        let name:String = weather
```
> ここの「:」の後のスペースの有無（Swiftでは基本的にvar hoge: Hoge = Hoge()の形）とか、まえ秋月さんにも言われてたけど気になってくるね〜

ここも

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -6,13 +6,40 @@
+    private func setWeatherImage() {
+        weather = YumemiWeather.fetchWeatherCondition()
+        let name:String = weather
+        var color:UIColor
```
同上

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -6,13 +6,40 @@
+        weather = YumemiWeather.fetchWeatherCondition()
+        let name:String = weather
+        var color:UIColor
+        if weather == "sunny" {
```
これはenum自分で用意してやった方が良さそうな感じするな〜

rikkkk1206 reviewed (changes requested):
まず色々コード上にコメント残したからそれらは対応して欲しいけど、
基本的な動作としては問題ないと思います。

以下は余裕あれば（というかここまで考慮できたらなお良しだねって感じ）だけど、せっかくだから考えてみて欲しい。

setWeatherImage() ← この名前を見た人は「weatherImageViewに画像をセットする」動作を想像すると思う。実際に動かしてもその通りの動作をするけど、さらに厳密に言うとグローバル変数であるweatherの値がこの変数内で更新されている。これは典型的な[副作用](https://zenn.dev/ueshun/articles/fcf1e27eca3cf5)の例だね。

まあ例によって今くらいの規模感なら問題なさそうに見えるけど、
例えば今後の変更で、
weatherを別のタイミングで更新する → 画像を更新したいからsetWeatherImage()を呼びだす → weatherは関数内で自動で更新されるので当然期待通りの挙動にならない
なんてことが起きるのは容易に想像できる。こうなると色んなパターンの悲劇が起こりうる。

てことで、今回はせっかくなのでこのくらい深く考えてもらって、対応し直して欲しい。
対応方法はいくつか思いつくけど、そこも考えてみて。

rikkkk1206 commented:
ちなみに↑とは直接は関係ないけど[CQS原則](https://qiita.com/pakkun/items/7dc5a9b6bc57225a3673)ってのがあるから、もし知らなかったなら一回読んでみて〜

KokiIto-45 committed:
session/2 天気予報を取得し、表示する

変数宣言の表記を修正
weatherの情報をグローバルで持たないように修正
Weatherクラスを作成し、sunny,cloudyなどのconditionをEnum型で保持、また各conditionごとのimageColorも同時に保持するように修正
また、setWeatherImage関数を、getWeatherConditionViaAPIとsetWeatherImageの二つに分離し、setWeatherImageはWeatehr型さえ用意すれば任意のタイミングで画像を変更できるように修正

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -12,32 +12,51 @@ class ViewController: UIViewController {
     
-    private var weather:String = ""
+    class Weather {
+        enum WeatherEnum: String {
+            case sunny = "sunny"
+            case cloudy = "cloudy"
+            case rain = "rainy"
+        }
```
実はSwiftのEnumにStringを継承させて[値型](https://qiita.com/hachinobu/items/392c96820588d1c03b0c#%E5%80%A4%E5%9E%8Benumraw-value-enum)にした場合、初期値を指定しなければcase名と同じ文字列をrawValueとして持たせてくれるんですね〜

KokiIto-45 replied:
@rikkkk1206 
なるほど、、！だとしたら冗長ですね、削除しておきます。

KokiIto-45 commented:
@rikkkk1206 
FBいただきありがとうございます。
まず変数宣言時の表記についてswiftの基本的な表記方法に則り変更しました。
次に、

> これはenum自分で用意してやった方が良さそうな感じするな〜

こちらについては、Weather型のclassを用意して、WeatherEnumと天気ごとのcolorを保持するようにしました。なのでsetImageView関数内も多少スッキリさせることができたのではないかと思います。ただclassのinit関数に条件分岐を移動させただけなので、もう少しうまい方法があるかもしれないです。

最後に副作用に関連した部分についてなのですが、これは指摘の通り、拡張性を考えた時にさまざまな問題が発生することは明らかだと思いました。
なので、今回自分がとった修正方針としては
- setImageView関数をgetWeatherConditionViaAPIとsetImageViewに分離
    -  getWeatherConditionViaAPIは[CQC原則](https://qiita.com/pakkun/items/7dc5a9b6bc57225a3673)に照らした際にクエリとして機能するようにした(APIから天気情報を取得してWeather型を返すのみ)
    - setImageViewでは、引数にWeather型を受け取りImageViewを変更するのみとした(CQC原則で考えた時にコマンドのように扱える。ただコマンドの順番を入れ替えても同じ挙動をする訳ではないので厳密には違うかもしれません。。)

以上のように修正してみました。もっとベターな方法はあると思うので、もっとこうした方がいい点などがあれば教えていただきたいです。

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -12,32 +12,51 @@ class ViewController: UIViewController {
+        
+        init(weatherString: String) {
+            self.condition = WeatherEnum(rawValue: weatherString)!
+            if self.condition == .sunny {
+                self.imageColor = .red
+            } else if self.condition == .cloudy {
+                self.imageColor = .gray
+            } else {
+                self.imageColor = .blue
+            }
```
ここは別の場所に括り出したいな〜
これもまたいくつか方法あるから、伊藤くんなりの方法でここの処理をすっきりさせてみてほしい

rikkkk1206 reviewed (changes requested):
全体的に非常に良くなったと思う👌
天気関連の情報をWeatherにまとめられているし、setWeatherImage()にweather情報を引数として渡すのを強制してたり、weatherを取得するための関数を用意してるのも🙆‍♂️

ただまだ少し気になる箇所にコメント残したから、それぞれ直してください

rikkkk1206 commented:
@KokiIto-45 
なんか行き違いになった！！
おれなりにいい方法は考えたけど、もうここに書いちゃっていい？？
それとももう少し自分で考える？

KokiIto-45 commented:
@rikkkk1206 

> おれなりにいい方法は考えたけど、もうここに書いちゃっていい？？
それとももう少し自分で考える？

行き違いになってしまってすみません！
imageColorをセットしている部分ですよね？もう少し自分で考えてみたいと思います！

KokiIto-45 committed:
session/2 API 天気予報を取得し、表示する

天気ごとのimageColorを設定するときに、WeatherEnumのindexを取得してweatherColorArrayからimageColorを設定するように修正

KokiIto-45 commented:
@rikkkk1206 
imageColorをセットしている箇所、修正してみました。が、本当はconditionと1対1で対応させるような構造にしたいと思って探したのですがいまいち良い方法が思いつかなかったので若干妥協案という感じです。中田さんのお考えもよければぜひ聞きたいです。

rikkkk1206 commented:
@KokiIto-45 
なるほどね、これはEnumの使い方に慣れてないと難しいかもね〜。
おれは↓みたいな感じがいいかなと思った。これが最適かはわからないけど。

```
class Weather {
        enum WeatherEnum: String {
            case sunny
            case cloudy
            case rainy
            
            func getColor() -> UIColor {
                switch self {
                case .sunny: return .red
                case .cloudy: return .gray
                case .rainy: return .blue
                }
            }
        }
        let condition: WeatherEnum
        let imageColor: UIColor
        
        init(weatherString: String) {
            self.condition = WeatherEnum(rawValue: weatherString)!
            self.imageColor = self.condition.getColor()
        }
    }
```

KokiIto-45 commented:
　@rikkkk1206 
なるほど、、enumの中に関数を書けるんですね。自分も断然こっちの方がいいと思います。
こちらを採用してまたpushさせていただきます！

KokiIto-45 committed:
session/2 天気予報を取得し、表示する

WeatherEnum型の中にgetColor関数を追加

rikkkk1206 commented on YumemiTutorial/ViewController.swift:
```diff
@@ -6,13 +6,62 @@
+                }
+            }
+        }
+        static let weatherColorArray:[UIColor] = [.red, .gray, .blue]
```
これ要らなそうだから消しといてね

rikkkk1206 reviewed (approved):
問題なさそう◎
一応コメント残したとこあるから、それ直したら自分でマージしちゃっていいよ〜

KokiIto-45 committed:
session/2 天気予報を取得し、表示する

余計なプロパティを削除
